source/common/network/dns_impl.cc

```diff
@@ -5,93 +5,127 @@
 #include "common/network/address_impl.h"
 #include "common/network/utility.h"
 
+#include "ares.h"
 #include "event2/event.h"
 
 namespace Network {
 
 DnsResolverImpl::DnsResolverImpl(Event::DispatcherImpl& dispatcher) : dispatcher_(dispatcher) {
-  // This sets us up to receive signals on an fd when async DNS resolved are completed.
-  sigset_t mask;
-  sigemptyset(&mask);
-  sigaddset(&mask, Event::Libevent::Global::DNS_SIGNAL_ID);
-  signal_fd_ = signalfd(-1, &mask, SFD_NONBLOCK);
-  RELEASE_ASSERT(-1 != signal_fd_);
-
-  event_assign(&signal_read_event_, &dispatcher_.base(), signal_fd_,
-               EV_READ | EV_PERSIST, [](evutil_socket_t, short, void* arg) -> void {
-                 static_cast<DnsResolverImpl*>(arg)->onSignal();
-               }, this);
-
-  event_add(&signal_read_event_, nullptr);
+  // This is also done in main(), to satisfy the requirement that c-ares is
+  // initialized prior to threading. The additional call to ares_library_init()
+  // here is a nop in normal execution, but exists for testing where we don't
+  // launch via main().
+  ares_library_init(ARES_LIB_INIT_ALL);
+  struct ares_options options;
```

Comment:
There are a lot of cases of this in the change, if you could remove them all that would be good.